### PR TITLE
Feat(tsc-elasticsearch)

### PR DESCRIPTION
**Pull request by Qinyouzeng**
# Elasticsearch Support for Observability（log and trace）

### CODE REVIEW

File: src/Contrib/StackSdks/Masa.Contrib.StackSdks.Tsc.Elasticseach/Const/MappingConst.cs [hunk 1]
```diff
@@ -0,0 +1,16 @@
+// Copyright (c) MASA Stack All rights reserved.
+// Licensed under the MIT License. See LICENSE.txt in the project root for license information.
+
+namespace Masa.Contrib.StackSdks.Tsc.Elasticseach.Const;
+
+internal static class MappingConst
```
Comment (zhenlei520) [inline src/Contrib/StackSdks/Masa.Contrib.StackSdks.Tsc.Elasticseach/Const/MappingConst.cs:6]: MappingConst -> MappingConstant

File: src/Contrib/StackSdks/Masa.Contrib.StackSdks.Tsc.Elasticseach/Const/ElasticConst.cs [hunk 2]
```diff
@@ -0,0 +1,40 @@
+// Copyright (c) MASA Stack All rights reserved.
+// Licensed under the MIT License. See LICENSE.txt in the project root for license information.
+
+namespace Masa.Contrib.StackSdks.Tsc.Elasticseach.Const;
```
Comment (zhenlei520) [inline src/Contrib/StackSdks/Masa.Contrib.StackSdks.Tsc.Elasticseach/Const/ElasticConst.cs:4]: Move under `Constants`

File: src/Contrib/StackSdks/Masa.Contrib.StackSdks.Tsc.Elasticseach/Extenistions/ServiceExtenistion.cs [hunk 3]
```diff
@@ -0,0 +1,115 @@
+// Copyright (c) MASA Stack All rights reserved.
+// Licensed under the MIT License. See LICENSE.txt in the project root for license information.
+
+namespace Masa.Contrib.StackSdks.Tsc.Elasticseach;
```
Comment (zhenlei520) [inline src/Contrib/StackSdks/Masa.Contrib.StackSdks.Tsc.Elasticseach/Extenistions/ServiceExtenistion.cs:4]: namespace Microsoft.Extensions.DependencyInjection;

File: src/Contrib/StackSdks/Masa.Contrib.StackSdks.Tsc.Elasticseach/LogService.cs [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+namespace Masa.Contrib.StackSdks.Tsc.Elasticseach;
+
+internal class LogService : ILogService
+{
+    private readonly IElasticClient _client;
+    private readonly ICallerFactory _callerFactory;
+
+    public LogService(IElasticsearchFactory elasticsearchFactory,ICallerFactory callerFactory)
+    {
```
Comment (zhenlei520) [inline src/Contrib/StackSdks/Masa.Contrib.StackSdks.Tsc.Elasticseach/LogService.cs:12]: format code

File: src/Contrib/StackSdks/Masa.Contrib.StackSdks.Tsc.Elasticseach/Masa.Contrib.StackSdks.Tsc.Elasticseach.csproj [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+
+  <ItemGroup>
+    <PackageReference Include="NEST" Version="$(NESTPackageVersion)" />
+  </ItemGroup>
+
+  <ItemGroup>
+    <ProjectReference Include="..\..\..\BuildingBlocks\StackSdks\Masa.BuildingBlocks.StackSdks.Tsc.Contracts\Masa.BuildingBlocks.StackSdks.Tsc.Contracts.csproj" />
+    <ProjectReference Include="..\..\..\Utils\Data\Masa.Utils.Data.Elasticsearch\Masa.Utils.Data.Elasticsearch.csproj" />
+    <ProjectReference Include="..\..\Service\Caller\Masa.Contrib.Service.Caller.HttpClient\Masa.Contrib.Service.Caller.HttpClient.csproj" />
+    <ProjectReference Include="..\..\Service\Caller\Masa.Contrib.Service.Caller\Masa.Contrib.Service.Caller.csproj" />
```
Comment (zhenlei520) [inline src/Contrib/StackSdks/Masa.Contrib.StackSdks.Tsc.Elasticseach/Masa.Contrib.StackSdks.Tsc.Elasticseach.csproj:17]: Remove Masa.Contrib.Service.Caller.csproj reference

File: src/Contrib/StackSdks/Masa.Contrib.StackSdks.Tsc.Elasticseach/Model/TraceResponseDtoExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+namespace Masa.Contrib.StackSdks.Tsc.Elasticseach.Model;
+
+internal static class TraceResponseDtoExtensions
+{
+    private static JsonSerializerOptions _serializerOptions = new JsonSerializerOptions()
+    {
+        PropertyNamingPolicy = JsonNamingPolicy.CamelCase
+    };
+
+    public static bool IsHttp(this ElasticseachTraceResponseDto data, out ElasticseachTraceHttpResponseDto result)
```
Comment (zhenlei520) [inline src/Contrib/StackSdks/Masa.Contrib.StackSdks.Tsc.Elasticseach/Model/TraceResponseDtoExtensions.cs:13]: inappropriate method name

File: src/Contrib/StackSdks/Masa.Contrib.StackSdks.Tsc.Elasticseach/Model/TraceResponseDtoExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+        {
+            return value.EnumerateArray().Select(item => item.ToString()).ToArray();
+        }
+        else
+        {
+            return new string[] { value.ToString() };
+        }
+    }
+
+    public static bool IsDatabase(this ElasticseachTraceResponseDto data, out ElasticseachTraceDatabaseResponseDto result)
```
Comment (zhenlei520) [inline src/Contrib/StackSdks/Masa.Contrib.StackSdks.Tsc.Elasticseach/Model/TraceResponseDtoExtensions.cs:42]: inappropriate method name

File: src/Contrib/StackSdks/Masa.Contrib.StackSdks.Tsc.Elasticseach/Model/TraceResponseDtoExtensions.cs [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+            return false;
+
+        result = data.Attributes.Deserialize<ElasticseachTraceExceptionResponseDto>();
+
+        return true;
+    }
+
+    private static bool ContainsKeies(this Dictionary<string, object> dic, IEnumerable<string> keys)
+    {
+        return dic != null && dic.Any() && dic.Keys.Any(key => keys.Any(k => key == k));
```
Comment (zhenlei520) [inline src/Contrib/StackSdks/Masa.Contrib.StackSdks.Tsc.Elasticseach/Model/TraceResponseDtoExtensions.cs:64]: Realize under modification

File: src/Contrib/StackSdks/Masa.Contrib.StackSdks.Tsc.Elasticseach/TraceService.cs [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+
+namespace Masa.Contrib.StackSdks.Tsc.Elasticseach;
+
+internal class TraceService : ITraceService
+{
+    private readonly IElasticClient _client;
+
+    public TraceService(IElasticsearchFactory elasticsearchFactory)
+    {
+        _client = elasticsearchFactory.CreateElasticClient();
```
Comment (zhenlei520) [inline src/Contrib/StackSdks/Masa.Contrib.StackSdks.Tsc.Elasticseach/TraceService.cs:12]: Use Specified ElasticClient

File: src/Contrib/StackSdks/Tests/Masa.Contrib.StackSdks.Tsc.Elasticseach.Tests/Extensions/ServiceExtenistionTests.cs [hunk 10]
```diff
@@ -0,0 +1,137 @@
+// Copyright (c) MASA Stack All rights reserved.
+// Licensed under the MIT License. See LICENSE.txt in the project root for license information.
+
+using Masa.Utils.Data.Elasticsearch;
```
Comment (zhenlei520) [inline src/Contrib/StackSdks/Tests/Masa.Contrib.StackSdks.Tsc.Elasticseach.Tests/Extensions/ServiceExtenistionTests.cs:4]: use global using

File: src/Contrib/StackSdks/Tests/Masa.Contrib.StackSdks.Tsc.Elasticseach.Tests/Masa.Contrib.StackSdks.Tsc.Elasticseach.Tests.csproj [hunk 11]
```diff
@@ -0,0 +1,30 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>net6.0</TargetFramework>
+    <ImplicitUsings>enable</ImplicitUsings>
+    <Nullable>enable</Nullable>
+
```
Comment (zhenlei520) [inline src/Contrib/StackSdks/Tests/Masa.Contrib.StackSdks.Tsc.Elasticseach.Tests/Masa.Contrib.StackSdks.Tsc.Elasticseach.Tests.csproj:7]: remove empty line

**Review by zhenlei520 — CHANGES_REQUESTED** [review]

File: src/BuildingBlocks/StackSdks/Masa.BuildingBlocks.StackSdks.Tsc.Contracts/Model/FieldOrderDto.cs [hunk 12]
```diff
@@ -0,0 +1,11 @@
+// Copyright (c) MASA Stack All rights reserved.
+// Licensed under the MIT License. See LICENSE.txt in the project root for license information.
+
+namespace Masa.BuildingBlocks.StackSdks.Tsc.Contracts.Model;
+
+public class FieldOrderDto
+{
+    public string Name { get; set; }
+
+    public bool? IsAsc { get; set; }
```
[the remark this answers is not in the document]
Comment (doddgu) [reply src/BuildingBlocks/StackSdks/Masa.BuildingBlocks.StackSdks.Tsc.Contracts/Model/FieldOrderDto.cs]: Default is desc.

File: src/BuildingBlocks/Data/Masa.BuildingBlocks.Data.Contracts/Request/RequestPageBase.cs [hunk 13]
```diff
@@ -0,0 +1,11 @@
+// Copyright (c) MASA Stack All rights reserved.
+// Licensed under the MIT License. See LICENSE.txt in the project root for license information.
+
+namespace Masa.BuildingBlocks.Data.Contracts.Request;
+
+public class RequestPageBase
+{
+    public int Page { get; set; } = 1;
+
+    public int Size { get; set; } = 20;
```
Comment (zhenlei520) [inline src/BuildingBlocks/Data/Masa.BuildingBlocks.Data.Contracts/Request/RequestPageBase.cs:10]: Page -> PageSize

File: src/BuildingBlocks/StackSdks/Masa.BuildingBlocks.StackSdk.Tsc/Model/Log/PaginationDto.cs [hunk 14]
```diff
@@ -0,0 +1,23 @@
+// Copyright (c) MASA Stack All rights reserved.
+// Licensed under the MIT License. See LICENSE.txt in the project root for license information.
+
+namespace Masa.BuildingBlocks.StackSdks.Tsc.Contracts.Model;
+
+public sealed class PaginationDto<TModel>
```
Comment (zhenlei520) [inline src/BuildingBlocks/StackSdks/Masa.BuildingBlocks.StackSdk.Tsc/Model/Log/PaginationDto.cs:6]: Should inherit PaginatedListBase<TEntity>

File: src/Contrib/StackSdks/Masa.Contrib.StackSdks.Tsc.Elasticseach/Extenistions/DictionaryExtenistions.cs [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+            var value = source[key];
+            var newKey = key[prefix.Length..];
+            if (convertFunc != null)
+                value = convertFunc(source[key]);
+            result.Add(newKey, (T)value!);
+        }
+        return result;
+    }
+
+    public static T ToObject<T>(this Dictionary<string, object> dic)
```
Comment (zhenlei520) [inline src/Contrib/StackSdks/Masa.Contrib.StackSdks.Tsc.Elasticseach/Extenistions/DictionaryExtenistions.cs:29]: ToObject -> ConvertTo

File: src/Contrib/StackSdks/Masa.Contrib.StackSdks.Tsc.Elasticseach/Extenistions/JsonElementExtensions.cs [hunk 16]
```diff
@@ -0,0 +1,73 @@
+// Copyright (c) MASA Stack All rights reserved.
+// Licensed under the MIT License. See LICENSE.txt in the project root for license information.
+[assembly: InternalsVisibleTo("Masa.Contrib.StackSdks.Tsc.Elasticseach.Tests")]
+namespace System;
+
+internal static class JsonElementExtensions
+{
+    public static IEnumerable<KeyValuePair<string, object>> ToKeyValuePairs(this JsonElement value)
```
Comment (zhenlei520) [inline src/Contrib/StackSdks/Masa.Contrib.StackSdks.Tsc.Elasticseach/Extenistions/JsonElementExtensions.cs:8]: ToKeyValuePairs -> ConvertToKeyValuePairs

File: src/Contrib/StackSdks/Masa.Contrib.StackSdks.Tsc.Elasticseach/Model/TraceResponseDtoExtensions.cs [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+            }
+            else
+            {
+                return new string[] { value.ToString() };
+            }
+        }
+        return new string[] { obj.ToString()! };
+    }
+
+    private static bool ContainsKeies(this Dictionary<string, object> dic, IEnumerable<string> keys)
```
Comment (zhenlei520) [inline src/Contrib/StackSdks/Masa.Contrib.StackSdks.Tsc.Elasticseach/Model/TraceResponseDtoExtensions.cs:59]: Method name does not match function

**Review by zhenlei520 — CHANGES_REQUESTED** [review]

File: src/Contrib/StackSdks/Masa.Contrib.StackSdks.Tsc.Elasticseach/Model/TraceResponseDtoExtensions.cs (hunk 17 above)
Comment (Qinyouzeng) [reply src/Contrib/StackSdks/Masa.Contrib.StackSdks.Tsc.Elasticseach/Model/TraceResponseDtoExtensions.cs]: This method is : if anyone value from `keys`  in the dictionary keys then is true

File: src/BuildingBlocks/StackSdks/Masa.BuildingBlocks.StackSdk.Tsc/Model/Log/PaginationDto.cs (hunk 14 above)
Comment (Qinyouzeng) [reply src/BuildingBlocks/StackSdks/Masa.BuildingBlocks.StackSdk.Tsc/Model/Log/PaginationDto.cs]: Is not used, has removed.

File: src/Contrib/StackSdks/Masa.Contrib.StackSdks.Tsc.Elasticseach/Model/TraceResponseDtoExtensions.cs [hunk 18]
```diff
@@ -0,0 +1,72 @@
...
+        }
+        return new string[] { obj.ToString()! };
+    }
+
+    private static bool IsContainsAnyKey(Dictionary<string, object> source, string[] keys)
+    {
+        if (source == null || !source.Any() || keys == null || !keys.Any())
+            return false;
+        if (keys.Length == 1)
+            return source.Keys.Any(key => key == keys[0]);
```
Comment (zhenlei520) [inline src/Contrib/StackSdks/Masa.Contrib.StackSdks.Tsc.Elasticseach/Model/TraceResponseDtoExtensions.cs:68]: source.ContainsKey(keys[0])

**Review by zhenlei520 — CHANGES_REQUESTED** [review]

**Review by zhenlei520 — APPROVED** [review]
@doddgu LGTM

File: src/BuildingBlocks/StackSdks/Masa.BuildingBlocks.StackSdks.Tsc.Contracts/Model/Trace/TraceDatabaseResponseDto.cs [hunk 19]
```diff
@@ -0,0 +1,35 @@
...
+    #region Cassandra
+
+    public virtual int CassandraPageSize { get; set; }
+    public virtual string CassandraConsistencyLevel { get; set; }
+    public virtual string CassandraTable { get; set; }
+    public virtual bool CassandraIdempotence { get; set; }
+    public virtual bool CassandraSpeculativeExecutionCount { get; set; }
+    public virtual string CassandraCoordinatorId { get; set; }
+    public virtual string CassandraCoordinatorDc { get; set; }
+    #endregion
```
Comment (doddgu) [inline src/BuildingBlocks/StackSdks/Masa.BuildingBlocks.StackSdks.Tsc.Contracts/Model/Trace/TraceDatabaseResponseDto.cs:34]: Add blank line.

**Review by doddgu — CHANGES_REQUESTED** [review]

**Review by doddgu — APPROVED** [review]
LGTM

**Pull request merged**